src/openms_gui/source/VISUAL/LayerData.cpp

```diff
@@ -374,4 +411,86 @@ namespace OpenMS
     if (annotations_changed) { hit.setPeakAnnotations(fas); }
   }
 
+  LayerAnnotatorBase::LayerAnnotatorBase(const FileTypes::FileTypeList& supported_types, const String& file_dialog_text)
+    : supported_types_(supported_types),
+      file_dialog_text_(file_dialog_text)
+  {
+  }
+
+  bool LayerAnnotatorBase::annotate(LayerData& layer, LogWindow* const log, const String& current_path) const
+  {
+    // warn if hidden layer => wrong layer selected...
+    if (!layer.visible)
+    {
+      log->appendNewHeader(LogWindow::LogState::NOTICE, "The current layer is not visible", "Have you selected the right layer for this action? Aborting.");
+      return false;
+    }
+
+    // load id data
+    QString fname = QFileDialog::getOpenFileName(nullptr,
+                                                 file_dialog_text_.toQString(),
+                                                 current_path.toQString(),
+                                                 supported_types_.toFileDialogFilter(FileTypes::Filter::BOTH, true).toQString());
+    if (fname.isEmpty()) return false;
+
+    FileTypes::Type type = FileHandler::getType(fname);
+
+    if (!supported_types_.contains(type))
+    {
+      QMessageBox::warning(nullptr, "Error", QString("Unsupported file type. No annotation performed."));
+      return false;
+    }
+
+    bool success = annotateWorker_(layer, fname, log);
+    
+    if (success) log->appendNewHeader(LogWindow::LogState::NOTICE, "Done", "Annotation finished. Open identification view to see results!");
+
+    return success;
+  }
+
+  bool LayerAnnotatorPeptideID::annotateWorker_(LayerData& layer, const String& filename, LogWindow* const log) const
+  {
+    FileTypes::Type type = FileHandler::getType(filename);
+    vector<PeptideIdentification> identifications;
+    vector<ProteinIdentification> protein_identifications;
+    String document_id;
+    if (type == FileTypes::MZIDENTML) MzIdentMLFile().load(filename, protein_identifications, identifications);
+    else IdXMLFile().load(filename, protein_identifications, identifications, document_id);
+
+    layer.annotate(identifications, protein_identifications);
+    return true;
+  }
+
+  bool LayerAnnotatorAMS::annotateWorker_(LayerData& layer, const String& filename, LogWindow* const log) const
+  {
+    FeatureMap fm;
+    FeatureXMLFile().load(filename, fm);
+
+    // last protein ID must be from AccurateMassSearch (it gets appended there)
+    String engine = "no protein identification section found";
+    if (fm.getProteinIdentifications().size() > 0)
+    {
+      engine = fm.getProteinIdentifications().back().getSearchEngine();
+      if (engine == "AccurateMassSearch")
```

Comment:
would be great if we could either get this string from the algorithm or from constants.h

Reply:
done